usage.c

```diff
@@ -6,17 +6,40 @@
 #include "git-compat-util.h"
 #include "cache.h"
 
+static void replace_control_chars(char *str, size_t size)
+{
+	size_t i;
+
+	for (i = 0; i < size; i++) {
+		if (iscntrl(str[i]) && str[i] != '\t' && str[i] != '\n')
+			str[i] = '?';
+	}
+}
+
+/* Atomically report (prefix + vsnprintf(err, params) + '\n') to stderr. */
 void vreportf(const char *prefix, const char *err, va_list params)
 {
 	char msg[4096];
-	char *p;
+	int ret;
+	size_t prefix_size, total_size;
 
-	vsnprintf(msg, sizeof(msg), err, params);
-	for (p = msg; *p; p++) {
-		if (iscntrl(*p) && *p != '\t' && *p != '\n')
-			*p = '?';
-	}
-	fprintf(stderr, "%s%s\n", prefix, msg);
+	prefix_size = strlen(prefix);
+	if (prefix_size >= sizeof(msg))
+		BUG("vreportf: prefix is too long");
+
+	memcpy(msg, prefix, prefix_size);
+
+	ret = vsnprintf(msg + prefix_size, sizeof(msg) - prefix_size, err, params);
+	if (ret < 0)
+		BUG("your vsnprintf is broken (returned %d)", ret);
+
+	total_size = strlen(msg);   /* vsnprintf() returns _desired_ size */
+	msg[total_size++] = '\n';   /* it's ok to overwrite terminating NULL */
+
+	replace_control_chars(msg, total_size);
+
+	fflush(stderr);             /* flush FILE* before writing lowlevel fd */
```

Comment:
This comment just repeats what the statement says, and is therefore non-DRY.

A more interesting bit of information is *why* this is necessary, and this information naturally should go into the commit message, not into the code, as code comments like ones talking about specifics of a certain platform are _prone_ to become stale, which we try to avoid in Git's source code. The reasoning behind a patch is pretty much tied to the commit itself, and should therefore be described in the commit message.

Speaking of commit messages: the style you chose deviates noticeably from the style found in the existing commit history, and I wonder what reasons there are for that.

Reply:
Can you suggest a better wording? It's necessary because FILE* and fd are different layers of abstraction. I think that problems can occur on any platform.

Reply:
As for commit message, what do you mean by style? Headers? I thought there's too much information to put in one stream, easier to navigate with headers. Just like in any other text.

Reply:
No, I cannot suggest a better wording, as my suggestion to remove the unnecessary comment still stands.

Reply:
I don't expect that an average reader will understand quickly why fflush() could be needed here, so sounds like exactly when comments are good?

How about this wording? `We're going to write to low-level fd while higher-level FILE* misght still have something cached`